components/notebook-controller/e2e/suite_test.go

```diff
@@ -0,0 +1,231 @@
+	envVar, exists := os.LookupEnv(variable)
+	if !exists {
+		return defaultVal
+	}
+	return envVar
+}
+
+// Setting up the ENV. This will run once before all the tests.
+var _ = SynchronizedBeforeSuite(func() []byte {
+	logf.SetLogger(zap.LoggerTo(GinkgoWriter, true))
+	By("Creating a k8s cluster.")
+	if !useExistingCluster {
+		err := bootstrapk8sCluster()
+		Expect(err).NotTo(HaveOccurred())
+		Expect(err).NotTo(HaveOccurred())
+	}
+	existingCluster := true
+	testEnv = &envtest.Environment{UseExistingCluster: &existingCluster}
+	var err error
+	cfg, err = testEnv.Start()
+	Expect(err).ToNot(HaveOccurred())
+	Expect(cfg).ToNot(BeNil())
+
+	err = nbv1beta1.AddToScheme(scheme.Scheme)
+	Expect(err).NotTo(HaveOccurred())
+
+	// Applying kustomizations from the config folder
+	err = applyKustomizations()
+	Expect(err).NotTo(HaveOccurred())
+
+	// +kubebuilder:scaffold:scheme
+
+	k8sClient, err = client.New(cfg, client.Options{Scheme: scheme.Scheme})
+	Expect(err).ToNot(HaveOccurred())
+	Expect(k8sClient).ToNot(BeNil())
+
+	return nil
+
+}, func(data []byte) {
+})
+
+// Cleaning the ENV
+var _ = SynchronizedAfterSuite(func() {
+}, func() {
+	if !useExistingCluster {
+		By("Tearing down the test k8s cluster created by the test.")
+		Expect(deleteK3DCluster(clusterName)).Should(BeNil())
+	} else {
+		By("Deleting the deployed CRD's in the existing cluster.")
+		// Deleting CRD's that were deployed as part of the test.
+		err := deleteKustomizations()
+		Expect(err).Should(BeNil())
+	}
+})
+
+// bootstrapk8sCluster creates a k3d cluster with a random name. Loads the docker image into the cluster.
+// The docker image name retrieved from the ENV variable TestImage.
+func bootstrapk8sCluster() error {
+	clusterName = fmt.Sprintf("e2e-notebook-kf-cluster-%s", randStringRunes(6))
+
+	cmd := exec.Command("k3d", "cluster", "create", clusterName, "-i", k3DockerImage,
+		"--no-lb", "--wait", "--timeout", "5m",
+		"--update-default-kubeconfig=false")
+	cmd.Stderr, cmd.Stdout = os.Stderr, os.Stdout
+	err := cmd.Run()
+
+	if err != nil {
+		return err
+	}
+	cmd = exec.Command("k3d", "kubeconfig", "write", clusterName)
+	err = cmd.Run()
+
+	if err != nil {
+		return err
+	}
+	kubeconfigPath := path.Join(os.Getenv("HOME"), fmt.Sprintf(".k3d/kubeconfig-%s.yaml", clusterName))
+	_ = os.Setenv("KUBECONFIG", kubeconfigPath)
+	imageName := os.Getenv(TestImage)
+	if len(imageName) == 0 {
+		return errors.New("the env variable TEST_IMAGE is not set to load the image into the k3d cluster")
+	}
+	cmd = exec.Command("k3d", "image", "import", "-c", clusterName, imageName)
+	cmd.Stderr, cmd.Stdout = os.Stderr, os.Stdout
+	return cmd.Run()
+}
+
+// deleteK3DCluster deletes the dynamically created k3d cluster.
+func deleteK3DCluster(clusterName string) error {
+	return exec.Command("k3d", "cluster", "delete", clusterName).Run()
+}
+
+// applyKustomizations applies kustomizations into the k8s clusters.
+func applyKustomizations() error {
```

Comment:
Should we make `applyKustomizations` take a list of paths as input, so it's re-usable?